diff -ru g10/keygen.c g10/keygen.c
--- g10/keygen.c	2011-01-10 19:09:46.000000000 +0100
+++ g10/keygen.c	2011-12-22 12:45:02.000000000 +0100
@@ -1774,14 +1774,20 @@
 static unsigned
 ask_keysize (int algo, unsigned int primary_keysize)
 {
-  unsigned int nbits, min, def = DEFAULT_STD_KEYSIZE, max=4096;
+  unsigned int nbits, min, def = DEFAULT_STD_KEYSIZE, max;
   int for_subkey = !!primary_keysize;
   int autocomp = 0;
 
   if(opt.expert)
-    min=512;
+    {
+      min=512;
+      max=8192;
+    }
   else
-    min=1024;
+    {
+      min=1024;
+      max=4096;
+    }
 
   if (primary_keysize && !opt.expert)
     {
