diff --git a/configure.ac b/configure.ac
index 0b365f3ab..4ea7156e1 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1928,7 +1928,7 @@ AC_DEFINE_UNQUOTED(TPM2DAEMON_SOCK_NAME, "S.tpm2daemon",
 AC_DEFINE_UNQUOTED(DIRMNGR_SOCK_NAME, "S.dirmngr",
                    [The name of the dirmngr socket])
 AC_DEFINE_UNQUOTED(DIRMNGR_DEFAULT_KEYSERVER,
-                   "hkps://keyserver.ubuntu.com",
+                   "hkps://pgpkeys.eu",
       [The default keyserver for dirmngr to use, if none is explicitly given])
 
 AC_DEFINE_UNQUOTED(GPGEXT_GPG, "gpg", [The standard binary file suffix])
diff --git a/dirmngr/server.c b/dirmngr/server.c
index 30c48fc51..7a8996455 100644
--- a/dirmngr/server.c
+++ b/dirmngr/server.c
@@ -2223,22 +2223,25 @@ make_keyserver_item (const char *uri, uri_item_t *r_item)
    */
   if (!strcmp (uri, "hkps://keys.gnupg.net")
       || !strcmp (uri, "keys.gnupg.net"))
-    uri = "hkps://keyserver.ubuntu.com";
+    uri = "hkps://pgpkeys.eu";
   else if (!strcmp (uri, "https://keys.gnupg.net"))
-    uri = "hkps://keyserver.ubuntu.com";
+    uri = "hkps://pgpkeys.eu";
   else if (!strcmp (uri, "hkp://keys.gnupg.net"))
-    uri = "hkp://keyserver.ubuntu.com";
+    uri = "hkps://pgpkeys.eu";
   else if (!strcmp (uri, "http://keys.gnupg.net"))
-    uri = "hkp://keyserver.ubuntu.com:80";
+    uri = "hkps://pgpkeys.eu";
   else if (!strcmp (uri, "hkps://http-keys.gnupg.net")
            || !strcmp (uri, "http-keys.gnupg.net"))
-    uri = "hkps://keyserver.ubuntu.com";
+    uri = "hkps://pgpkeys.eu";
   else if (!strcmp (uri, "https://http-keys.gnupg.net"))
-    uri = "hkps://keyserver.ubuntu.com";
+    uri = "hkps://pgpkeys.eu";
   else if (!strcmp (uri, "hkp://http-keys.gnupg.net"))
-    uri = "hkp://keyserver.ubuntu.com";
+    uri = "hkps://pgpkeys.eu";
   else if (!strcmp (uri, "http://http-keys.gnupg.net"))
-    uri = "hkp://keyserver.ubuntu.com:80";
+    uri = "hkps://pgpkeys.eu";
+  else if (!strcmp (uri, "hkps://keyserver.ubuntu.com")
+           || !strcmp (uri, "keyserver.ubuntu.com"))
+    uri = "hkps://pgpkeys.eu/";
 
   return ks_action_parse_uri (uri, r_item);
 }
