diff --git a/dirmngr/dirmngr.c b/dirmngr/dirmngr.c
index c3cdd56a7..50351df8f 100644
--- a/dirmngr/dirmngr.c
+++ b/dirmngr/dirmngr.c
@@ -1540,6 +1540,20 @@ main (int argc, char **argv)
                  DIRMNGR_INFO_NAME, socket_name, (ulong) pid);
       w32_ack_to_frontend ();
 #else
+#ifdef __APPLE__
+      /* Load macOS system certificates before fork() to avoid a
+         crash if Foundation methods are first used after fork.
+         For details see: http://www.sealiesoftware.com/blog/archive/2017/6/5/Objective-C_and_fork_in_macOS_1013.html
+
+         The system certificates are added to each gnutls session,
+         but the session is only established after the fork and thus
+         a call to get_system_certificate_authorities () would crash.
+
+         By calling get_system_certificate_authorities () before fork
+         the certificate authorities are already cached when the gnutls
+         session is configured.  */
+      get_system_certificate_authorities (0);
+#endif
       pid = fork();
       if (pid == (pid_t)-1)
         {
@@ -1729,6 +1743,7 @@ main (int argc, char **argv)
 static void
 cleanup (void)
 {
+  clear_system_certificate_authorities ();
   crl_cache_deinit ();
   cert_cache_deinit (1);
   reload_dns_stuff (1);
diff --git a/dirmngr/http-common.c b/dirmngr/http-common.c
index 3b6cd44c7..2ba55e06f 100644
--- a/dirmngr/http-common.c
+++ b/dirmngr/http-common.c
@@ -23,6 +23,10 @@
 #include <stdlib.h>
 #include <string.h>
 
+#ifdef __APPLE__
+# include <Security/Security.h>
+#endif
+
 #include "dirmngr.h"
 #include "http-common.h"
 
@@ -45,6 +49,112 @@ get_default_keyserver (int name_only)
       result += 3;
       /* Assert that there is no port given.  */
       log_assert (!strchr (result, ':'));
-    }
+  }
   return result;
 }
+
+CFMutableArrayRef
+get_system_certificate_authorities (int clear)
+{
+#ifdef __APPLE__
+  static CFMutableArrayRef certificate_authorities = NULL;
+
+  if (certificate_authorities != NULL || clear == 1)
+    {
+      if (certificate_authorities != NULL && clear == 1)
+        {
+          log_debug ("clearing system certificate authorities cache\n");
+          CFRelease (certificate_authorities);
+          certificate_authorities = NULL;
+          log_debug ("system certificate authorities cache: %p\n",
+                     certificate_authorities);
+        }
+
+      return certificate_authorities;
+    }
+
+  log_debug ("load system certificate authorities into the cache\n");
+  certificate_authorities = CFArrayCreateMutable (NULL, 0,
+                                                  &kCFTypeArrayCallBacks);
+
+  SecTrustSettingsDomain domain[] =
+    {
+      kSecTrustSettingsDomainUser,
+      kSecTrustSettingsDomainAdmin,
+      kSecTrustSettingsDomainSystem
+    };
+
+  for (size_t d = 0; d < sizeof (domain) / sizeof (*domain); d++)
+    {
+      CFArrayRef certs = NULL;
+      OSStatus status = SecTrustSettingsCopyCertificates (domain[d], &certs);
+      if (status != errSecSuccess)
+        continue;
+
+      CFIndex cert_count = CFArrayGetCount (certs);
+      for (CFIndex i = 0; i < cert_count; i++)
+        {
+          SecCertificateRef cert = (void *)CFArrayGetValueAtIndex (certs, i);
+          CFDataRef der = NULL;
+
+          status = SecItemExport (cert, kSecFormatX509Cert, 0, NULL, &der);
+          if (status != errSecSuccess)
+            {
+              if (der)
+                CFRelease (der);
+              continue;
+            }
+
+          CFArrayAppendValue (certificate_authorities, der);
+        }
+
+      CFRelease (certs);
+    }
+
+  log_info ("system certificate authorities caching complete - found %ld certificate authorities\n",
+            (long)CFArrayGetCount (certificate_authorities));
+
+  return certificate_authorities;
+#else
+  return NULL;
+#endif
+}
+
+#ifdef HTTP_USE_GNUTLS
+int
+add_system_certificate_authorities (gnutls_certificate_credentials_t cred)
+{
+#ifdef __APPLE__
+  CFArrayRef certificate_authorities = get_system_certificate_authorities (0);
+  CFIndex ca_count = CFArrayGetCount (certificate_authorities);
+
+  log_debug ("gnutls: adding %ld system certificate authorities\n",
+             (long)ca_count);
+
+  int valid_cas = 0;
+  for (CFIndex i = 0; i < ca_count; i++)
+    {
+      CFDataRef cert = CFArrayGetValueAtIndex (certificate_authorities, i);
+      gnutls_datum_t datum =
+        {
+          .data = (unsigned char *)CFDataGetBytePtr (cert),
+          .size = CFDataGetLength (cert),
+        };
+      int rc = gnutls_certificate_set_x509_trust_mem (cred, &datum,
+                                                      GNUTLS_X509_FMT_DER);
+      if (rc > 0)
+        valid_cas += rc;
+    }
+
+  return valid_cas > 0 ? valid_cas : GNUTLS_E_FILE_ERROR;
+#else
+  return gnutls_certificate_set_x509_system_trust (cred);
+#endif
+}
+#endif
+
+void
+clear_system_certificate_authorities (void)
+{
+  (void)get_system_certificate_authorities (1);
+}
diff --git a/dirmngr/http-common.h b/dirmngr/http-common.h
index ddb340de6..cbc68aa43 100644
--- a/dirmngr/http-common.h
+++ b/dirmngr/http-common.h
@@ -20,8 +20,21 @@
 #ifndef HTTP_COMMON_H
 #define HTTP_COMMON_H
 
+#ifdef __APPLE__
+# include <CoreFoundation/CoreFoundation.h>
+#endif
+
+#ifdef HTTP_USE_GNUTLS
+# include <gnutls/gnutls.h>
+# include <gnutls/x509.h>
+#endif /*HTTP_USE_GNUTLS*/
+
 const char *get_default_keyserver (int name_only);
 
+CFMutableArrayRef get_system_certificate_authorities (int clear);
+int add_system_certificate_authorities (gnutls_certificate_credentials_t cred);
+void clear_system_certificate_authorities (void);
+
 void http_reinitialize (void);
 
 #endif /* HTTP_COMMON_H */
diff --git a/dirmngr/http.c b/dirmngr/http.c
index 5723a13f2..cec66a0bc 100644
--- a/dirmngr/http.c
+++ b/dirmngr/http.c
@@ -957,7 +957,7 @@ http_session_new (http_session_t *r_session,
       {
         static int shown;
 
-        rc = gnutls_certificate_set_x509_system_trust (sess->certcred);
+        rc = add_system_certificate_authorities (sess->certcred);
         if (rc < 0)
           log_info ("setting system CAs failed: %s\n", gnutls_strerror (rc));
         else if (!shown)
