# Also install gpg-zip.


--- a/tools/Makefile.in
+++ b/tools/Makefile.in
@@ -653,8 +653,8 @@ libcommontlsnpth = ../common/libcommontlsnpth.a
 AM_CFLAGS = $(LIBGCRYPT_CFLAGS) $(GPG_ERROR_CFLAGS) $(LIBASSUAN_CFLAGS)
 sbin_SCRIPTS = addgnupghome applygnupgdefaults
 
-# bin_SCRIPTS += gpg-zip
-@HAVE_USTAR_TRUE@noinst_SCRIPTS = gpg-zip
+sbin_SCRIPTS += gpg-zip
+# @HAVE_USTAR_TRUE@noinst_SCRIPTS = gpg-zip
 @BUILD_WKS_TOOLS_FALSE@gpg_wks_server = 
 @BUILD_WKS_TOOLS_TRUE@gpg_wks_server = gpg-wks-server
 common_libs = $(libcommon)
