--- a/lib/system/certs.c
+++ b/lib/system/certs.c
@@ -101,7 +101,7 @@ int _gnutls_find_config_path(char *path, size_t max_size)
 	return 0;
 }
 
-#if defined(DEFAULT_TRUST_STORE_FILE) || (defined(DEFAULT_TRUST_STORE_PKCS11) && defined(ENABLE_PKCS11))
+#if (defined(DEFAULT_TRUST_STORE_FILE) && (!defined(__APPLE__) || defined(__APPLE__) && __MAC_OS_X_VERSION_MIN_REQUIRED < 1070)) || (defined(DEFAULT_TRUST_STORE_PKCS11) && defined(ENABLE_PKCS11))
 static
 int
 add_system_trust(gnutls_x509_trust_list_t list,

--- a/lib/system/certs.c
+++ b/lib/system/certs.c
@@ -291,6 +291,7 @@ static
 int add_system_trust(gnutls_x509_trust_list_t list, unsigned int tl_flags,
 		     unsigned int tl_vflags)
 {
+	int ret;
 	int r=0;
 
 	SecTrustSettingsDomain domain[] = { kSecTrustSettingsDomainUser,
@@ -333,6 +334,17 @@ int add_system_trust(gnutls_x509_trust_list_t list, unsigned int tl_flags,
 		CFRelease(certs);
 	}
 
+#ifdef DEFAULT_TRUST_STORE_FILE
+	ret =
+	    gnutls_x509_trust_list_add_trust_file(list,
+						  DEFAULT_TRUST_STORE_FILE,
+						  crl_file,
+						  GNUTLS_X509_FMT_PEM,
+						  tl_flags, tl_vflags);
+	if (ret > 0)
+		r += ret;
+#endif
+
 #ifdef DEFAULT_BLACKLIST_FILE
 	ret = gnutls_x509_trust_list_remove_trust_file(list, DEFAULT_BLACKLIST_FILE, GNUTLS_X509_FMT_PEM);
 	if (ret < 0) {
